src/Compilers/CSharp/Portable/SourceGeneration/IndentingStringBuilder.cs

```diff
@@ -0,0 +1,497 @@
+    /// is at the start of a line.
+    /// </summary>
+    private void AppendSingleLine(ReadOnlySpan<char> line, string? originalLine, bool skipIndent)
+    {
+        if (line.Length == 0)
+            return;
+
+        var builder = this.Builder;
+
+        if (!skipIndent && !IsEndOfLineCharacter(line[0]))
+        {
+            if (builder.Length == 0 || IsEndOfLineCharacter(builder[^1]))
+                builder.Append(_currentIndentation);
+        }
+
+#if NET
+        builder.Append(line);
+#else
+        if (originalLine != null)
+        {
+            builder.Append(originalLine);
+        }
+        else
+        {
+            foreach (var ch in line)
+                builder.Append(ch);
+        }
+#endif
+    }
+
+    private static bool IsEndOfLineCharacter(char ch)
+        => EndOfLineCharacters.IndexOf(ch) >= 0;
+
+    /// <summary>
+    /// Writes content to the underlying buffer.  If the buffer is at the start of a line, then indentation will be
+    /// appended first before the content.  By default, for performance reasons, the content is assumed to contain no
+    /// end of line characters in it.  If the content may contain end of line characters, then <see langword="true"/>
+    /// should be passed in for <paramref name="splitContent"/>.  This will cause the provided content to be split into
+    /// constituent lines, with each line being appended one at a time.
+    /// </summary>
+    /// <param name="skipIndent">If true, will not indent content, even if starting a new line.</param>
+    public IndentingStringBuilder Write(string content, bool splitContent = false, bool skipIndent = false)
+        => Write(content.AsSpan(), content, splitContent, skipIndent);
+
+    /// <inheritdoc cref="Write(string, bool, bool)"/>
+    public IndentingStringBuilder Write(ReadOnlySpan<char> content, bool splitContent = false, bool skipIndent = false)
+        => Write(content, originalString: null, splitContent, skipIndent);
+
+    /// <inheritdoc cref="Write(string, bool, bool)"/>
+    private IndentingStringBuilder Write(ReadOnlySpan<char> content, string? originalString, bool splitContent, bool skipIndent = false)
+    {
+        if (splitContent)
+        {
+            while (content.Length > 0)
+            {
+                var endOfLineIndex = content.IndexOfAny(EndOfLineCharacters);
+                if (endOfLineIndex < 0)
+                {
+                    // no new line, append the rest of the content to the buffer.
+                    AppendSingleLine(content, originalLine: null, skipIndent);
+                }
+                else
+                {
+                    while (endOfLineIndex < content.Length & IsEndOfLineCharacter(content[endOfLineIndex + 1]))
+                        endOfLineIndex++;
+
+                    AppendSingleLine(content[0..endOfLineIndex], originalLine: null, skipIndent);
```

Comment:
📝 This line needs to be moved _before_ the `while` loop (or `if` loop if you switch). Otherwise, the `\r` of a `\r\n` sequence will be passed on to `AppendSingleLine`